gap.go

```diff
@@ -311,94 +365,152 @@ func (buf *rawAdvertisementPayload) addFromOptions(options AdvertisementOptions)
 	// legacy advertising) because of the 31-byte advertisement packet
 	// limit.
 	for _, uuid := range options.ServiceUUIDs {
-		if !buf.addServiceUUID(uuid) {
-			return false
+		if err := buf.addServiceUUID(uuid); err != nil {
+			return err
 		}
 	}
 
 	for _, element := range options.ManufacturerData {
-		if !buf.addManufacturerData(element.CompanyID, element.Data) {
-			return false
+		if err := buf.addManufacturerData(element.UUID, element.Data); err != nil {
+			return err
+		}
+	}
+
+	for _, element := range options.ServiceData {
+		if err := buf.addServiceData(element.UUID, element.Data); err != nil {
+			return err
 		}
 	}
 
-	return true
+	return nil
 }
 
 // addManufacturerData adds manufacturer data ([]byte) entries to the advertisement payload.
-func (buf *rawAdvertisementPayload) addManufacturerData(key uint16, value []byte) (ok bool) {
+func (buf *rawAdvertisementPayload) addManufacturerData(uuid UUID, data []byte) (err error) {
 	// Check whether the field can fit this manufacturer data.
-	fieldLength := len(value) + 4
+	fieldLength := 1 + 1 + 2 + len(data) // 1 byte length, 1 byte ad type, 2 bytes uuid, actual manufacturer data
 	if int(buf.len)+fieldLength > len(buf.data) {
-		return false
+		return errAdvertisementPacketTooBig
 	}
 
 	// Add the data.
-	buf.data[buf.len+0] = uint8(fieldLength - 1)
-	buf.data[buf.len+1] = 0xff
-	buf.data[buf.len+2] = uint8(key)
-	buf.data[buf.len+3] = uint8(key >> 8)
-	copy(buf.data[buf.len+4:], value)
+	buf.data[buf.len+0] = byte(fieldLength - 1)
+	buf.data[buf.len+1] = 0xFF
+	buf.data[buf.len+2] = byte(uuid.Get16Bit())
+	buf.data[buf.len+3] = byte(uuid.Get16Bit() >> 8)
+	copy(buf.data[buf.len+4:], data)
```

Comment:
With all these extra (unrelated!) changes it becomes a bit hard to review this PR. I appreciate the effort but I'd prefer to keep PRs small and focused on a single thing.